This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [1]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = 'yxTNuEFm8j5mza_uYS3A'

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

### Import Modules

In [2]:
# First, import the relevant modules
import requests
import collections

### Access API and pull daily sample

In [3]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned

stock_code = 'AFX_X'
column_start = '0'
start_date = '2019-01-01'
end_date = '2019-01-02'

#url =   'https://www.quandl.com/api/v3/datasets/WIKI/AFX_X.json?column_index=4&start_date=2018-01-01&end_date=2018-01-02&collapse=monthly+&transform=rdiff&api_key='+API_KEY
#url =   'https://www.quandl.com/api/v3/datasets/WIKI/'+stock_code+'.json?column_index='+column_start+'&start_date='+start_date+'&end_date='+end_date+'&collapse='+'monthly'+'&transform='+'rdiff'+'&api_key='+API_KEY

url = 'https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?&start_date=2019-01-01&end_date=2019-01-02&api_key='+API_KEY

r = requests.get(url)

### Inspect JSON structure

In [4]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure

r.status_code

200

In [5]:
#r.headers

In [6]:
#r.text

In [7]:
json_data = r.json()
#json_data

In [8]:
for key, value in json_data.items():
    print(key)
    print()
    print()
    for k,v in value.items():
        print(k)
        print(v)
        print()

dataset


id
10095370

dataset_code
AFX_X

database_code
FSE

name
Carl Zeiss Meditec (AFX_X)

description
Stock Prices for Carl Zeiss Meditec (2020-07-20) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704

refreshed_at
2020-08-03T22:16:16.600Z

newest_available_date
2020-08-03

oldest_available_date
2000-06-07

column_names
['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']

frequency
daily

type
Time Series

premium
False

limit
None

transform
None

column_index
None

start_date
2019-01-01

end_date
2019-01-02

data
[['2019-01-02', None, 70.85, 68.25, 69.85, None, None, None, None, None, None]]

collapse
None

order
None

database_id
6129



### Task List

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

### Task 1

In [9]:
url2 = 'https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?&start_date=2017-01-01&end_date=2017-12-31&api_key='+API_KEY
r2 = requests.get(url2)

No data returned for 01/01/2017 or 30/12/2017 to 31/12/2017. 

### Task 2

In [10]:
json = r2.json()
#json

### Task 3

This section has bit of exploratory piece before getting to answer to question

Below code is to identify column names

In [11]:
column_names = json['dataset']['column_names']
column_names

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

Below code is to look at what data is available

In [12]:
data = json['dataset']['data']
#data

#data0 = json['dataset']['data'][0]

Below code is to create new dictionary combining column names and data

In [13]:
from collections import defaultdict
from collections import OrderedDict
new_dict2 = OrderedDict()

#dict(zip(column_names, data[0]))

for x in data:
    temp_dict = dict(zip(column_names, x))
    new_dict2[temp_dict['Date']] = temp_dict

#new_dict2
#type(new_dict2)

Below code is to view output from dictionary to see if it made sense

In [14]:
#list(new_dict2.items())[0:2]

In [15]:
#for date, info in new_dict2.items():
    #print(date, ': ', new_dict2[date]['Close'])

Below code is used to answer question of highest and lowest opening prices throughou the year.

Worked of the assumption we did not require to know the date of these hit but if this was required, alternative solution would be neeed. Also, unsure if solution around 'None' values is correct.

In [16]:
open_price2 = []

for date, info in new_dict2.items():
    if new_dict2[date]['Open'] == None:
        continue
    else:
        open_price2.append(new_dict2[date]['Open'])

#open_price2

print(max(open_price2))
print(min(open_price2))

53.11
34.0


### Task 4

Code is produced similar to previous task and continues throughout project. If incorrect assumptions, all need to be updated.

In [17]:
high_to_low = []

for date, info in new_dict2.items():
    if new_dict2[date]['High'] == None:
        continue
    elif new_dict2[date]['Low'] == None:
        continue
    else:
        change = new_dict2[date]['High'] - new_dict2[date]['Low']
        high_to_low.append(change)

print(max(high_to_low, key=abs))

2.8100000000000023


### Task 5

Code below has two loops to calculate answer. First is simlar to above tasks. 

Second assumes that start by comparing Day 2 to Day 1 and so on until you reach maximum number of days in list and then breaks loop.

In [18]:
close_price = []

for date, info in new_dict2.items():
    if new_dict2[date]['Close'] == None:
        continue 
        #Maybe should put in default value but this would skew results
    else:
        close = new_dict2[date]['Close']
        close_price.append(close)
        
day_to_day = []
i = 0
range_close = len(close_price)

for day in close_price:
    i = i+1
    if i >= range_close:
        break
    else:
        change = close_price[i] - day
        day_to_day.append(change)

print(max(day_to_day, key=abs))

2.559999999999995


### Task 6

Use sum and lenght of trading volume list to calculate mean.

In [19]:
trading_volume = []

for date, info in new_dict2.items():
    if new_dict2[date]['Traded Volume'] == None:
        continue 
    else:
        volume = new_dict2[date]['Traded Volume']
        trading_volume.append(volume)

sum_volume = sum(trading_volume)
len_volume = len(trading_volume)
mean_volume = sum_volume / len_volume

print(mean_volume)

89124.33725490196


### Task 7

Found middle value in length of list.

Then sorted listed so in order.

Then found indexes either side of middle and calculated average for them.

In [20]:
middle = len_volume / 2
#middle

In [21]:
trading_volume.sort()
trading_volume

median_1 = trading_volume[127]
median_2 = trading_volume[128]

median = (median_2 + median_1) / 2
median

76443.0

### Answers to tasks (where possible)

1) Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).

2) Convert the returned JSON object into a Python dictionary.

3) Calculate what the highest and lowest opening prices were for the stock in this period.
Highest = 53.11
Lowest = 34.0

4) What was the largest change in any one day (based on High and Low price)?
2.81

5) What was the largest change between any two days (based on Closing Price)?
2.56

6) What was the average daily trading volume during this year?
89,124

7) (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)
76,443
